tests/tests/test_mender_shell.py

```diff
@@ -0,0 +1,91 @@
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+import logging
+import pytest
+
+from testutils.api import proto_shell
+from testutils.api.proto_shell import WRONG_BEHAVIOR
+from ..common_setup import standard_setup_one_client_bootstrapped
+from ..MenderAPI import devconnect
+
+if WRONG_BEHAVIOR:
+    logger = logging.getLogger("websockets")
+    logger.setLevel(logging.DEBUG)
+    logger.addHandler(logging.StreamHandler())
+
+
+class TestMenderShell:
+    def test_regular_protocol_commands(self, standard_setup_one_client_bootstrapped):
+        device = standard_setup_one_client_bootstrapped.device
+
+        if WRONG_BEHAVIOR:
+            # MEN-4273.
+            device.run(
+                """sed -i -e '2i"SkipVerify": true,' /etc/mender/mender-shell.conf"""
+            )
+            device.run("systemctl restart mender-shell")
+
+        with devconnect.get_websocket() as ws:
+            # Start shell.
+            shell = proto_shell.ProtoShell(ws)
+            shell.startShell()
+
+            # Drain any initial output from the prompt. It should end in either "# "
+            # (root) or "$ " (user).
+            output = shell.recvOutput()
+            assert output[-2:].decode() in [
+                "# ",
+                "$ ",
+            ], "Could not detect shell prompt."
+
+            if not WRONG_BEHAVIOR:
+                # Starting the shell again should be a no-op. It should return that it
+                # is already started, as long as the shell limit is 1.
+                # MEN-4240.
+                shell.startShell()
+
+            # Test if a simple command works.
+            shell.sendInput("ls /\n".encode())
+            output = shell.recvOutput()
+            output = output.decode()
+            assert "usr" in output
+            assert "etc" in output
+
+            # Try to stop shell.
+            shell.stopShell()
+
+            # Make sure we can not send anything to the shell.
+            shell.sendInput("ls /\n".encode())
+            output = shell.recvOutput()
+            output = output.decode()
+            assert "usr" not in output
+            assert "etc" not in output
+
+            # Start it again.
+            shell.startShell()
+
+            if not WRONG_BEHAVIOR:
+                # Drain any initial output from the prompt. It should end in either "# "
+                # (root) or "$ " (user).
+                # MEN-4240 or MEN-4141, possibly a combination.
+                output = shell.recvOutput()
+                assert output[-2:].decode() in [
+                    "# ",
+                    "$ ",
+                ], "Could not detect shell prompt."
```

Comment:
Maybe refactor this in a separate method, so that it can be reused with the above code.

Reply:
Maybe.. Since it has side effects (draining the output pipe), I think I prefer to keep this one inline for clarity. The Black formatter is making this look unreasonably big, in reality it is only two lines.